server/routes/charts.js

```diff
@@ -189,4 +189,28 @@ router
       return res.status(500).json({ message: error.message })
     }
   })
+  .patch(ensureAuthenticated, async (req, res) => {
+    try {
+      const { dataDb } = req.app.locals
+      const { chart_id } = req.params
+      const sourceChart = await dataDb.collection(collections.charts).findOne(
+        {
+          _id: new ObjectID(chart_id),
+        },
+        { projection: { _id: 0 } }
+      )
+      const { insertedId } = await dataDb
+        .collection(collections.charts)
+        .insertOne({
+          ...sourceChart,
+          title: sourceChart.title + ' Duplicate',
+        })
+
+      if (insertedId) return res.status(200).json({ data: insertedId })
+      return res.status(404).json({ message: 'Chart was not duplicated' })
```

Comment:
should we return a 404 not found if we can't find the source chart and some other error (maybe 500) if we can't save the duplicate?
```suggestion
      const sourceChart = await dataDb.collection(collections.charts).findOne(
        {
          _id: new ObjectID(chart_id),
        },
        { projection: { _id: 0 } }
      )
      if(!sourceChart.id) return res.status(404).json({ message: `Chart ${chart_id} was not found` })
      
      const { insertedId } = await dataDb
        .collection(collections.charts)
        .insertOne({
          ...sourceChart,
          title: sourceChart.title + ' Duplicate',
        })

      if (insertedId) return res.status(200).json({ data: insertedId })
      return res.status(500).json({ message: 'Chart could not be saved' })

```

Reply:
The catch part of the function should be able to take care of the rejected promise.

Reply:
right 😵‍💫 my snippet makes no sense 🤦 

We have two promises in the same try/catch,
- the first finds the existing document by id, if that fails we could return a "404 Not Found" (though I think currently this would land us in the catch (500))
- the second saves a new document to the db, if that fails we could return "500 Could not be saved"
  - as you pointed out this second error is handed by `catch` ✅ 
  - to return the 404 from the first promise, I think we'd need a separate try/catch around the `sourceChart` query 🟠

Reply:
Could use your help here, because all of the errors resolve to an instance of Error instead of MongoError. I could change the 500 to a 404 and then an arbitrary error message is sent to the UI.

Reply:
Sure, let me know if you want to huddle. 
I think how you have it is fine, though, if I'm understanding this right we'll never hit line 210 because if the insert call fails we'll go to the catch block immediately, skipping 210, so we'd be returning a 500 instead of 400 -- which is totally cool. If we want to return different errors based on which query failed, we could break this out into separate try/catches so we can return different errors based on one query failing vs the other, but that may be more trouble than it's worth.
```diff
+   try {
+     const { dataDb } = req.app.locals
+     const { chart_id } = req.params
+     let sourceChart
-   try {
-     const { dataDb } = req.app.locals
-     const { chart_id } = req.params
-     const sourceChart = await dataDb.collection(collections.charts).findOne(
+     sourceChart = await dataDb.collection(collections.charts).findOne(
        {
          _id: new ObjectID(chart_id),
        },
        { projection: { _id: 0 } }
      )
+   } catch (_error) {
+      return res.status(404).json({ message: `Chart ${chart_id} not found` })
+   }

+   try {
      const { insertedId } = await dataDb
        .collection(collections.charts)
        .insertOne({
          ...sourceChart,
          title: sourceChart.title + ' Duplicate',
        })

-      if (insertedId) return res.status(200).json({ data: insertedId })
+      return res.status(200).json({ data: insertedId })

-      return res.status(404).json({ message: 'Chart was not duplicated' }) // <-- line 210
    } catch (error) {
      return res.status(500).json({ message: error.message })
    }
  })
```

Reply:
https://github.com/AMP-SCZ/dpdash/pull/127/commits/b5f38cb418e1f1c0da0dcafb7354b98d8c00c1d0